pkg/driver/trace_driver.go

```diff
@@ -307,46 +237,46 @@ func (d *Driver) functionsDriver(list *list.List, announceFunctionDone *sync.Wai
 
 		previousIATSum += iat.Microseconds()
 
+		if !d.Configuration.TestMode {
+			waitForInvocations.Add(1)
+
+			go d.invokeFunction(&InvocationMetadata{
+				RootFunction:          list,
+				Phase:                 currentPhase,
+				MinuteIndex:           minuteIndex,
+				InvocationIndex:       invocationIndex,
+				SuccessCount:          &successfulInvocations,
+				FailedCount:           &failedInvocations,
+				FailedCountByMinute:   failedInvocationByMinute,
+				RecordOutputChannel:   recordOutputChannel,
+				AnnounceDoneWG:        &waitForInvocations,
+				AnnounceDoneExe:       addInvocationsToGroup,
+				ReadOpenWhiskMetadata: readOpenWhiskMetadata,
+			}, iatIndex)
+		} else {
+			// To be used from within the Golang testing framework
+			log.Debugf("Test mode invocation fired.\n")
+
+			recordOutputChannel <- &mc.ExecutionRecord{
+				ExecutionRecordBase: mc.ExecutionRecordBase{
+					Phase:        int(currentPhase),
+					InvocationID: composeInvocationID(d.Configuration.TraceGranularity, minuteIndex, invocationIndex),
+					StartTime:    time.Now().UnixNano(),
+				},
+			}
+
+			successfulInvocations++
+		}
+		numberOfIssuedInvocations++
+		invocationIndex++
+
 		if function.InvocationStats.Invocations[minuteIndex] == invocationIndex || hasMinuteExpired(startOfMinute) {
 			readyToBreak := d.proceedToNextMinute(function, &minuteIndex, &invocationIndex, &startOfMinute,
```

Comment:
With current IAT structure, we might not call the proceedToNextMinute on the minute border (e.g., when the IAT=8s in RPS mode). The logic there should be updated to accommodate that.

I noticed that previousIATSum is updated there. Since now we have a flat IAT structure, we might have a problem that this IAT sum doesn't correspond to the sum of elements of IAT array up to i-th one.

Reply:
Have a look at the comment above.